### PR TITLE
[DPE-3136][DPE-3137] Add dpe-workflows & poetry

### PR DESCRIPTION
**Pull request by phvalguima**
This PR moves the opensearch-operator to use data-platform-workflows repo, it adds more behaviors to backups, as described below; and extends its integration testing.

Moving to data-platform-workflows brings opensearch to the same level in its testing infrastructure as other operators. It also enables testing to easily consume secrets from Github, as well as other features such as redact-secrets or support for self-hosted runners. This move covers [DPE-3136](https://warthogs.atlassian.net/browse/DPE-3136) and [DPE-3137](https://warthogs.atlassian.net/browse/DPE-3137). Now, opensearch-operator supports Poetry and can use poetry.lock to pin dependencies.

Important movements, besides changing to dpe-workflows:
* Moves methods that are repeated several times: `reset_restart_delay`, `c_writes`, `c_writes_runner` and `c_balanced_writes_runner` to a single place: `tests/integration/ha/conftest.py`
* Removes `tests/integration/conftest.py` as the build process is now executed by the DPE workflows
* Removes the root `tests/conftest.py`, as it was used for secrets management between 2.9 and 3.x, besides the new workflow uses relative paths starting in `tests/unit` and `tests/integration` folders
* Moves the tests declared in `tests/integration/relations/opensearch_provider/` to `tests/integration/relations` for the sake of simplifying relative paths
* Adds a local build of `application_charm` in `tests/integration/relations` folder

The backup test has been removed and will be readded in [DPE-2999 PR](https://github.com/canonical/opensearch-operator/pull/181).

[DPE-3136]: https://warthogs.atlassian.net/browse/DPE-3136?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-3137]: https://warthogs.atlassian.net/browse/DPE-3137?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by phvalguima** [comment]
The check is failing: `app: opensearch blocked -- message: 1 or more 'replica' shards are not assigned, please scale your application up.` and timing out
It is not related to this change.
Full check here: https://github.com/canonical/opensearch-operator/actions/runs/7940046438/job/21680946956?pr=178

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -21,6 +21,41 @@ this operator.
...
+tox -e build-production
+```
+
+### Build With Cache
+
+First, ensure you have the right dependencies:
+* charmcraft v2.5.4+
+* charmcraftcache
+
+By running the following commands:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply CONTRIBUTING.md:46]: Keeping the `build-production` in a subsection "Build Without Cache", as this represents our final build that will be released, right

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -32,41 +67,67 @@ source venv/bin/activate
...
-tox -e client-integration     # Validates the `opensearch-client` integration
-tox -e ha-integration         # HA tests
-tox -e h-scaling-integration  # HA tests specific to horizontal scaling
+tox -e integration -- tests/integration/ha/test_storage.py
 ```
 
-If you're running tests on juju 3, run the following command to change libjuju to the correct version:
+#### Running different versions of libjuju
+
+To try different versions of libjuju, run:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply CONTRIBUTING.md:88]: Ack, added more text explaining that point

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -32,41 +67,67 @@ source venv/bin/activate
...
 
 ```shell
-charmcraft pack
+tox -e integration -- tests/integration/ha/test_backups.py --group='microceph' # test backup service
+```
+
+And against public clouds + microceph:
+
+```shell
+SECRETS_FROM_GITHUB=$(cat <path-to>/credentials.json) tox -e integration -- tests/integration/ha/test_backups.py
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply CONTRIBUTING.md:108]: Ack, but that is needed for us to test locally. I added more text explaining this is development only and should be done only on your local machine

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+          fail-build: ${{ github.event_name == 'pull_request' }}
+
+  build:
+    name: Build charm
+    strategy:
+      fail-fast: false
+      matrix:
+        paths:
+        - to-charm-directory: "."
+        - to-charm-directory: "./tests/integration/relations/opensearch_provider/application-charm/"
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply .github/workflows/ci.yaml]: @carlcsaposs-canonical I am getting an error on the build whenever I remove the double-quotes. Readded it resolves the problem: https://github.com/canonical/opensearch-operator/actions/runs/7971145010/job/21760232295

File: pyproject.toml [hunk 5]
```diff
@@ -1,7 +1,92 @@
...
+codespell = "^2.2.5"
+shellcheck-py = "^0.9.0.5"
+
+[tool.poetry.group.unit.dependencies]
+juju = "^3.2.2"
+pytest = "^7.4.0"
+pytest-asyncio = "<0.23"
+pytest-mock = "^3.11.1"
+coverage = {extras = ["toml"], version = "^7.4.1"}
+parameterized = "^0.9.0"
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply pyproject.toml:70]: @carlcsaposs-canonical this is where I use it: https://github.com/canonical/opensearch-operator/blob/main/tests/unit/lib/test_backups.py#L133

File: tests/unit/lib/test_opensearch_secrets.py [hunk 6]
```diff
@@ -17,7 +16,11 @@
 from charm import OpenSearchOperatorCharm
 
 
-@pytest.mark.usefixtures("only_with_juju_secrets")
+class JujuVersionMock:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/unit/lib/test_opensearch_secrets.py:19]: @carlcsaposs-canonical the main challenge for us here is that we are not using `pytest` for unit testing everywhere. So, when I set it up like that, I get:
```
_________________________________________________________________________________ ERROR at setup of TestOpenSearchInternalData.test_data_has_0_app _________________________________________________________________________________
test_data_has_0_app does not support fixtures, maybe unittest.TestCase subclass?
Node id: tests/unit/lib/test_opensearch_secrets.py::TestOpenSearchInternalData::test_data_has_0_app
Function type: TestCaseFunction
```

I will file a bug to move our unit test framework to pytest universally.

[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/unit/lib/test_opensearch_secrets.py:19]: Filed an issue to look closer at it later: https://github.com/canonical/opensearch-operator/issues/188

File: tox.ini [hunk 7]
```diff
@@ -1,147 +1,108 @@
...
-    ha-storage: pytest -v --tb native --log-cli-level=INFO -s {posargs} {[vars]tests_path}/integration/ha/test_storage.py
-    ha-networking: pytest -v --tb native --log-cli-level=INFO -s {posargs} {[vars]tests_path}/integration/ha/test_ha_networking.py
-    ha-multi-clusters: pytest -v --tb native --log-cli-level=INFO -s {posargs} {[vars]tests_path}/integration/ha/test_ha_multi_clusters.py
-    large-deployments: pytest -v --tb native --log-cli-level=INFO -s {posargs} {[vars]tests_path}/integration/ha/test_large_deployments.py
-    plugins: pytest -v --tb native --log-cli-level=INFO -s {posargs} {[vars]tests_path}/integration/plugins/test_plugins.py
+set_env =
+    {[testenv]set_env}
+    # Workaround for https://github.com/python-poetry/poetry/issues/6958
+    POETRY_INSTALLER_PARALLEL = false
+    LIBJUJU_VERSION_SPECIFIER = {env:LIBJUJU_VERSION_SPECIFIER:3.2.3.0}
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tox.ini:84]: We are not targeting juju v2 anymore. It is safe to remove this.

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply CONTRIBUTING.md:102]: @carlcsaposs-canonical my recommendation is to always run tests locally as well. That makes our own lives easier for testing. Indeed, microceph is not installing because of that: https://pastebin.ubuntu.com/p/9JjmVQ6G9G/

File: CONTRIBUTING.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply CONTRIBUTING.md]: Ack, but I am not mentioning it in the docs, explicitly.

File: pyproject.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply pyproject.toml:67]: So, apologies, that was added to `test_opensearch_internal_data`. I will add it to the bug about moving away from `unittests` to `pytest`, but we cannot remove this dependency atm.

File: tests/unit/lib/test_opensearch_secrets.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/unit/lib/test_opensearch_secrets.py:19]: hi @carlcsaposs-canonical yes, I can imagine. I am just not getting it to fully work with some of the unit tests. I think this deserves a better look into, including to remove the `parametrized` dependency as well and unify under pytest. Let's break this out as another scope.

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply CONTRIBUTING.md:102]: Sorry, not "to always run" but to always "have the option". In any case, this is a discussion that we can take offline.

File: tests/integration/relations/conftest.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    if (
+        # This is a build process that should be only valid whenever we are not in CI
+        ("CI" not in os.environ or os.environ["CI"] != "true")
+        and (
+            # Now, check for the .charm file and if we had any recent updates in the sources
+            not glob.glob(f"{charmpath}/*.charm")
+            or should_rebuild_charm(glob.glob(f"{charmpath}/*.charm")[0])
+        )
+    ):
+        name = await reload(pytest_operator.plugin).OpsTest.build_charm(ops_test, charmpath)
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/relations/conftest.py]: hi @carlcsaposs-canonical yes, it is a bit hacky. I cannot get the workflow to correctly build my application_charm and then, locate that charm binary. I've simplified it here: https://github.com/canonical/opensearch-operator/actions/runs/7973599873/job/21768028749?pr=178

You can see the test fails to find the charm file.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -32,41 +59,64 @@ source venv/bin/activate
...
 ```shell
-export LIBJUJU_VERSION_SPECIFIER="==3.3.0.0"
+poetry add --lock --group integration juju@<YOUR CHOSEN VERSION>
 ```
 
-## Build charm
+#### FOR DEVELOPMENT ONLY: Testing Backups In Your Local Machine
+
+Backup testing installs microceph and, optionally, can run on AWS and GCP object stores.
+For that, pass the access / secret / service account information as env. variables.
```
Comment (Mehdi-Bendriss) [inline CONTRIBUTING.md:91]:
```suggestion
To setup your environment, you should set the: access / secret / service account information as environment variables.
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -32,41 +59,64 @@ source venv/bin/activate
...
 
 ```shell
-charmcraft pack
+SECRETS_FROM_GITHUB=$(cat <path-to>/credentials.json) tox -e integration -- tests/integration/ha/test_backups.py
 ```
 
-### Deploy
+Where, for AWS only, `credentials.json` should look like (the `IGNORE` is necessary to remove GCP):
+```python
+{ "AWS_ACCESS_KEY": ..., "AWS_SECRET_KEY": ..., "GCP_SECRET_KEY": "IGNORE" }
```
Comment (Mehdi-Bendriss) [inline CONTRIBUTING.md:107]: I'm a bit confused. Previously in the text it was said that those should be defined as environment variables. This now uses a json object.

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply CONTRIBUTING.md:102]: It would make sense to stay consistent - given that this PR is about moving to DP workflows, we should aim to have the same UX as the rest.

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -32,41 +59,64 @@ source venv/bin/activate
...
+```
+
+For GCP only, some additional information is needed. The pytest script runs with `boto3` and needs `GCP_{ACCESS,SECRET}_KEY` to clean up the object store; whereas OpenSearch uses the `service account` json file.
```
Comment (Mehdi-Bendriss) [inline CONTRIBUTING.md:110]: I don't think we support GCS at this stage. We should probably drop it.

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -32,41 +59,64 @@ source venv/bin/activate
...
+
+```python
+{ "AWS_SECRET_KEY": "IGNORE", "GCP_ACCESS_KEY": ..., "GCP_SECRET_KEY": ..., "GCP_SERVICE_ACCOUNT": ... }
```
Comment (Mehdi-Bendriss) [inline CONTRIBUTING.md:113]: I'm not a fan of the `"IGNORE"`, we shouldn't have to include other GCP related entries for AWS and vice-versa

File: tests/integration/ha/helpers.py [hunk 13]
```diff
@@ -476,8 +492,8 @@ async def wait_restore_finish(ops_test, unit_ip):
                         raise Exception()
 
 
-async def continuous_writes_increases(ops_test: OpsTest, unit_ip: str, app: str) -> bool:
-    """Asserts that TEST_BACKUP_INDEX is writable while under continuous writes.
+async def start_and_check_continuous_writes(ops_test: OpsTest, unit_ip: str, app: str) -> bool:
+    """Asserts that ContinuousWrites.INDEX_NAME is writable.
```
Comment (Mehdi-Bendriss) [inline tests/integration/ha/helpers.py:496]:
```suggestion
    """Start continuous writes and check that documents are increasing after some time.
```

File: tests/integration/relations/test_opensearch_provider.py [hunk 14]
```diff
@@ -452,72 +462,8 @@ async def test_admin_relation(ops_test: OpsTest):
     assert set(artists) == {"Herbie Hancock", "Lydian Collective", "Vulfpeck"}
 
 
+@pytest.mark.group(1)
 @pytest.mark.abort_on_fail
-@pytest.mark.usefixtures("only_without_juju_secrets")
-async def test_admin_permissions(ops_test: OpsTest):
```
Comment (Mehdi-Bendriss) [inline tests/integration/relations/test_opensearch_provider.py:457]: is this test **exactly identical** to the following one?

File: tests/integration/relations/test_opensearch_provider.py [hunk 15]
```diff
@@ -584,66 +530,8 @@ async def test_admin_permissions_secrets(ops_test: OpsTest):
         assert "Error:" in results[0], results
 
 
+@pytest.mark.group(1)
 @pytest.mark.abort_on_fail
-@pytest.mark.usefixtures("only_without_juju_secrets")
-async def test_normal_user_permissions(ops_test: OpsTest):
```
Comment (Mehdi-Bendriss) [inline tests/integration/relations/test_opensearch_provider.py:589]: is this test **exactly identical** to the following one?

File: .github/workflows/release.yaml [hunk 16]
```diff
@@ -8,48 +8,41 @@ on:
...
-      fail-fast: false
-      matrix:
-        include:
-          - snap: 3.1/stable
-            libjuju: ==3.3.0.0
-    uses: ./.github/workflows/integration.yaml
-    with:
-      juju-snap-channel: ${{ matrix.snap }}
-      libjuju-version-specifier: ${{ matrix.libjuju }}
+  # release-libraries:
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply .github/workflows/release.yaml:16]: @Mehdi-Bendriss `lib-check` is now done automatically in the workflow itself. This is *releasing libs*, which we are not currently doing via CI. I'd rather uncomment this code piece on a separate PR, wdyt?

File: tests/integration/ha/test_ha_networking.py [hunk 17]
```diff
@@ -146,11 +126,7 @@ async def test_full_network_cut_with_ip_change_node_with_elected_cm(
         ops_test, first_elected_cm_unit_ip, retries=3
     ), "Connection still possible to the first CM node where the network was cut."
 
-    # verify new writes are continuing by counting the number of writes before and after 5 seconds
-    writes = await c_writes.count()
-    time.sleep(5)
-    more_writes = await c_writes.count()
-    assert more_writes > writes, "Writes not continuing to DB"
+    assert_continuous_writes_increasing(c_writes)
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/ha/test_ha_networking.py:129]: Yes, that used to be "sync", thanks!

File: tests/integration/ha/helpers.py [hunk 18]
```diff
@@ -490,18 +506,19 @@ async def continuous_writes_increases(ops_test: OpsTest, unit_ip: str, app: str)
...
-async def backup_cluster(ops_test: OpsTest, leader_id: int) -> bool:
+async def backup_cluster(ops_test: OpsTest, leader_id: int) -> int:
     """Runs the backup of the cluster."""
     action = await run_action(ops_test, leader_id, "create-backup")
     logger.debug(f"create-backup output: {action}")
 
     await wait_backup_finish(ops_test, leader_id)
-    return action.status == "completed"
+    assert action.status == "completed"
+    return int(action.response["backup-id"])
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/ha/helpers.py:521]: Moved that assert up. "completed" here does not mean "finished backup"

File: tests/integration/ha/test_horizontal_scaling.py [hunk 19]
```diff
@@ -352,6 +336,8 @@ async def test_safe_scale_down_remove_leaders(
     app = (await app_name(ops_test)) or APP_NAME
     init_units_count = len(ops_test.model.applications[app].units)
 
+    c_writes_obj = c_writes
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/ha/test_horizontal_scaling.py:339]: That used to be a "wait", before I've found out pytest-asyncio mode was not correct

File: tests/integration/relations/conftest.py [hunk 20]
```diff
@@ -21,5 +21,5 @@ async def application_charm(ops_test: OpsTest):
 
 @pytest.fixture(scope="module")
 async def opensearch_charm(ops_test: OpsTest):
-    """Build the opensearch charm."""
+    """Build the opensearch charm as well."""
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/relations/conftest.py:24]: Meant for the "application_charm", we are following the same process

File: tests/integration/relations/test_opensearch_provider.py [hunk 21]
```diff
@@ -89,8 +91,11 @@ async def test_create_relation(ops_test: OpsTest, application_charm, opensearch_
         timeout=1600,
         status="active",
     )
+    # Return all update-status-hook-interval to 5m, as all idle_period values are set to 70s
+    await ops_test.model.set_config({"update-status-hook-interval": "5m"})
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/relations/test_opensearch_provider.py:95]: hi @Mehdi-Bendriss but then, we have to set all idle_period to less than 1m (e.g. 50s). Otherwise, juju will check status and you will see the latest refresh data being always smaller than 70s (where update-status runs at 60s)

File: tests/integration/relations/test_opensearch_provider.py (hunk 15 above)
Comment (phvalguima) [reply tests/integration/relations/test_opensearch_provider.py:589]: @Mehdi-Bendriss not sure I follow. This is only run when we do not have juju secrets. In our CIs, that is 100% of the cases

File: tests/integration/relations/test_opensearch_provider.py (hunk 14 above)
Comment (phvalguima) [reply tests/integration/relations/test_opensearch_provider.py:457]: @Mehdi-Bendriss not sure I follow. This is only run when we do not have juju secrets. In our CIs, that is 100% of the cases

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply CONTRIBUTING.md:112]: Ack, going to resolve this comment. It was also due to some settings left in tox.ini. I will clean them up, if not already done.

File: tests/integration/ha/test_storage.py [hunk 22]
```diff
@@ -4,98 +4,27 @@
...
+@pytest.mark.group(1)
 @pytest.mark.abort_on_fail
 @pytest.mark.skip_if_deployed
 async def test_build_and_deploy(ops_test: OpsTest) -> None:
     """Build and deploy one unit of OpenSearch."""
-    # it is possible for users to provide their own cluster for HA testing.
-    # Hence, check if there is a pre-existing cluster.
-    if await app_name(ops_test):
-        return
-
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/ha/test_storage.py:95]: The main target is the CI now, with the new workflow. I added it back, but I doubt we will be using it.

File: tests/integration/ha/helpers.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/ha/helpers.py:521]: Moved the action.status above. "completed" here does not mean "finished" and these assert vs. wait_for_backup have nothing to do with each other as checks.

File: tests/integration/ha/test_ha_networking.py [hunk 23]
```diff
@@ -35,40 +33,21 @@
...
-
+@pytest.mark.group(1)
 @pytest.mark.abort_on_fail
 @pytest.mark.skip_if_deployed
 async def test_build_and_deploy(ops_test: OpsTest) -> None:
     """Build and deploy one unit of OpenSearch."""
-    # it is possible for users to provide their own cluster for HA testing.
-    # Hence, check if there is a pre-existing cluster.
-    if await app_name(ops_test):
-        return
```
[the remark this answers is not in the document]
Comment (phvalguima) [reply tests/integration/ha/test_ha_networking.py:67]: The new workflow expects the CI, this scenario likely will never happen.

File: tests/integration/relations/test_opensearch_provider.py (hunk 14 above)
Comment (phvalguima) [reply tests/integration/relations/test_opensearch_provider.py:457]: They are 1-2-1 replicas. I will keep this comment open but close the remaining ones.

File: tests/integration/ha/test_storage.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply tests/integration/ha/test_storage.py]: Not sure I understand, our integration tests can run against an existing cluster as opposed to spinning up a new one every time we run the tests. 
This is what this does.

File: .github/workflows/release.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (Mehdi-Bendriss) [reply .github/workflows/release.yaml:16]: Ok, but please create a ticket with the full description and attach it here

[the remark this answers is not in the document]
Comment (phvalguima) [reply .github/workflows/release.yaml:16]: @Mehdi-Bendriss here: https://warthogs.atlassian.net/browse/DPE-3696

File: tests/integration/ha/helpers.py [hunk 24]
```diff
@@ -490,18 +504,19 @@ async def continuous_writes_increases(ops_test: OpsTest, unit_ip: str, app: str)
...
+    time.sleep(10)
     result = await writer.stop()
     return result.count > initial_count
 
 
-async def backup_cluster(ops_test: OpsTest, leader_id: int) -> bool:
+async def backup_cluster(ops_test: OpsTest, leader_id: int) -> int:
     """Runs the backup of the cluster."""
     action = await run_action(ops_test, leader_id, "create-backup")
+    assert action.status == "completed"
```
Comment (Mehdi-Bendriss) [inline tests/integration/ha/helpers.py:515]: Similar to a previous comment (wait_for_backup), it is discouraged to have asserts in non-assert methods, the preferred way is to throw a regular exception. There needs to be a separation between test functions and helpers. 
This can be the object of another PR.

File: tests/integration/ha/test_backups.py
Comment (Mehdi-Bendriss) [inline tests/integration/ha/test_backups.py:1]: This will be put back in another PR, I suppose?

**Review by Mehdi-Bendriss — APPROVED** [review]
Thanks Pedro. I left 1 comments that can be addressed in another PR

File: tests/integration/ha/helpers.py (hunk 24 above)
Comment (phvalguima) [reply tests/integration/ha/helpers.py:515]: @Mehdi-Bendriss done on this PR: https://github.com/canonical/opensearch-operator/blob/53d44ff91f894c48395c5c584bbc751b45f50b6b/tests/integration/ha/helpers.py

Which is based in this here.
